bazel_tools/client_server/runner_with_port_file/Main.hs

```diff
@@ -0,0 +1,38 @@
+-- Copyright (c) 2021 Digital Asset (Switzerland) GmbH and/or its affiliates. All rights reserved.
+-- SPDX-License-Identifier: Apache-2.0
+
+module Main (main) where
+
+import Data.List.Extra (replace, splitOn, stripInfix)
+import Data.Maybe (isJust)
+import System.Directory (removeFile)
+import System.Environment (getArgs)
+import System.Process (callProcess, proc, withCreateProcess)
+import System.IO.Temp (emptyTempFile, withSystemTempDirectory)
+
+import DA.PortFile
+
+main :: IO ()
+main = do
+  [clientExe, clientArgs, serverExe, serverArgs, _] <- getArgs
+  let splitArgs = filter (/= "") . splitOn " "
+  let splitServerArgs = splitArgs serverArgs
+  withSystemTempDirectory "runner" $ \tempDir -> do
+    (portFile, interpolatedServerArgs) <-
+      if any (isJust . stripInfix "%PORT_FILE%") splitServerArgs
+        then do
+          portFile <- emptyTempFile tempDir "port"
+          removeFile portFile
+          let interpolatedArgs = map (replace "%PORT_FILE%" portFile) splitServerArgs
+          return (Just portFile, interpolatedArgs)
+        else
+          return (Nothing, splitServerArgs)
+    let serverProc = proc serverExe interpolatedServerArgs
+    withCreateProcess serverProc $ \_stdin _stdout _stderr _ph -> do
+      maybePort <- mapM (readPortFile maxRetries) portFile
+      let splitClientArgs = splitArgs clientArgs
+      let interpolatedClientArgs =
+            case maybePort of
+              Nothing -> splitClientArgs
+              Just port -> map (replace "%PORT%" (show port)) splitClientArgs
```

Comment:
just always doing that now that you dropped the wait seems like it might simplify things a bit

Reply:
Dropped the wait, sure, but we still have no guarantee that the server actually wrote a port file.

I kept this as a `Maybe` because it allows us to use one runner everywhere. Do you think we're better off only using it where we definitely have a port file?

Reply:
Before your PR we have two types of client_server runners:

1. Based on port files which waits for the server to write a port file.
2. Based on hardcoded ports which tests the connection until the server is up.

The client needs to know the port the server is listening on so if you don’t get it from the port file you don’t have a lot of other options than hardcoding it in some form.

I understood this PR as introducing a variation of 1 where the port file is specified differently meaning it doesn’t make sense to use it if the server does not write a port file. It sounds like you have something different in mind but I think I don’t quite understand what.

Reply:
My end goal is that we have one runner. I can see an argument for one (everything is done through Bazel configuration), and an argument for two (two different ways of starting the server).

Which would you prefer?

Reply:
Definitely in favor of one runner but that doesn’t really answer the question I clearly failed to ask properly :)

We currently have waiting for a port file and waiting for a connection. You’re adding a third option in here where we wait for nothing. What do we need this for? In my ideal world everything runs on ephemeral ports for which we need the port file (otherwise we can’t tell the client tell what to connect to). Where that’s not possible, we hardcode and wait for the connection. I don’t see the need for the third option.

Reply:
I think I was getting ahead of myself. For now I'll simplify, though I think this logic will need to come back if we merge this runner with `runner_with_port_check`.